### DISCUSSION TITLE
Rails feature tests have stopped running javascript

### DISCUSSION THREAD

**Discussion by IanVaughan** [community/community#25683] [Actions]
<p>Having been happily using GitHub Actions to run my Rails/Rspec/Capybara tests for many months, it has now just stopped working!</p>
<p>The pipelines started failing on builds on 26th Aug, around 1200 GMT.<br>
Eg clicking on a link that performs a POST, now does a GET.<br>
(Using <a href="https://apidock.com/rails/ActionView/Helpers/UrlHelper/link_to" rel="nofollow noopener">https://apidock.com/rails/ActionView/Helpers/UrlHelper/link_to</a> “method: :post”)</p>
<p>Any test that requires JavaScript to run fails, eg :</p>
<pre><code class="lang-auto">    Failure/Error:
      page.accept_alert 'Are you sure?' do
        click_button('Create new application')
      end

    Capybara::ModalNotFound:
      Unable to find modal dialog with Are you sure?
    # ------------------
    # --- Caused by: ---
    # Selenium::WebDriver::Error::TimeoutError:
    #   timed out after 20 seconds (no such alert
    #     (Session info: headless chrome=84.0.4147.125))
</code></pre>
<p>The problem can be isolated from our code by looking at previous merges to our repo.<br>
Older merges to master that passed at the time are now failing on re-runs.<br>
It looks as if something changed in the base build on github actions?</p>
<p>Our config:</p>
<pre><code class="lang-ruby">name: "Tests"
on:
  push:

jobs:
  rspec:
    name: RSpec
    runs-on: ubuntu-latest
    env:
      RAILS_MASTER_KEY: ${{ secrets.RAILS_MASTER_KEY }}
      RAILS_ENV: test
      DATABASE_URL: postgres://postgres:postgres@localhost:5432/test
      REDIS_URL: redis://localhost:6379/0

    services:
      postgres:
        image: postgres:latest
        ports:
          - 5432:5432
        env:
          POSTGRES_PASSWORD: postgres
        options: &gt;-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

      redis:
        image: redis:latest
        ports: ["6379:6379"]
        options: --entrypoint redis-server

    steps:
      - uses: actions/checkout@v1

      - name: Setup ruby
        uses: ruby/setup-ruby@v1
        with:
          bundler-cache: true

      - name: Find yarn cache location
        id: yarn-cache
        run: echo "::set-output name=dir::$(yarn cache dir)"

      - name: JS package cache
        uses: actions/cache@v1
        with:
          path: ${{ steps.yarn-cache.outputs.dir }}
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-

      - name: Install packages
        run: |
          yarn install --pure-lockfile

      - name: Install postgres client
        run: sudo apt-get install libpq-dev

      - name: Install dependencies
        run: |
          gem install bundler
          bundle install --path vendor/bundle --jobs 4 --retry 3

      - name: Create database
        run: |
          bundler exec rails db:create
          bundler exec rails db:migrate

      - name: Yarn
        run: yarn test

      - name: Run tests
        run: bundler exec rspec
</code></pre>

**Comment by NicoNBW** [comment]
<p>Thanks for your feedback.</p>
<aside class="quote no-group" data-username="IanVaughan" data-post="1" data-topic="129485">
<div class="title">
<div class="quote-controls"></div>
 IanVaughan:</div>
<blockquote>
<p>The pipelines started failing on builds on 26th Aug, around 1200 GMT.</p>
</blockquote>
</aside>
<p>Do you mean it worked normally before, but now it fails? If so, did you make any specific changes in your code?<br>
Do you have a <strong>public repo</strong> demonstrating this? We can take a look to see what happened!</p>

**Comment by IanVaughan** [comment]
<aside class="quote no-group" data-username="niconbw" data-post="2" data-topic="129485">
<div class="title">
<div class="quote-controls"></div>
 niconbw:</div>
<blockquote>
<p>Do you mean it worked normally before, but now it fails?</p>
</blockquote>
</aside>
<p>Yes, it was working and now it’s failing, but, as I say, fails for previously passing (unchanged) builds.</p>
<aside class="quote no-group" data-username="niconbw" data-post="2" data-topic="129485">
<div class="title">
<div class="quote-controls"></div>
 niconbw:</div>
<blockquote>
<p>If so, did you make any specific changes in your code?</p>
</blockquote>
</aside>
<p>As I say above, previous merges to master that passed at the time are now failing on re-runs.<br>
In case that is not clear, a build that passed fine a week ago, which then we merged because it passed CI, now fails. That is, if I open that commit and see it shows a green tick stating the CI had passed, I open that and re-run the tests, they now fail. Thus meaning that it is not something under our control but something that GitHub have done to the image.</p>
<aside class="quote no-group" data-username="niconbw" data-post="2" data-topic="129485">
<div class="title">
<div class="quote-controls"></div>
 niconbw:</div>
<blockquote>
<p>Do you have a <strong>public repo</strong> demonstrating this? We can take a look to see what happened!</p>
</blockquote>
</aside>
<p>Unfortunately not this is a private repo, but unless you are an employee of GitHub, I don’t see what looking at the repo will show you, because as I say, no changes in our repo have caused this change.</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/ianvaughan">@IanVaughan</a><br>
Since it is a private repo, could you please create another public demo repo which can reproduce this issue so that we can investigate further.</p>

**Comment by IanVaughan** [comment]
<p>The issue has solved itself!</p>
<p>At a guess, a gem update has helped, but the fact that no gem update was made to cause the issue still concerns me.</p>
<p>Sorry I was unable to find time to create a public repo, but thanks for your help in any case. Ian</p>
<p>FWIW changes since raising this report:</p>
<pre><code class="lang-auto">&gt; git --no-pager diff --no-prefix -U0 HEAD 'HEAD@{2020-08-28 00:00:00}' -- Gemfile.lock
</code></pre>
<pre><code class="lang-diff">-    aws-partitions (1.376.0)
-    aws-sdk-core (3.108.0)
+    aws-partitions (1.359.0)
+    aws-sdk-core (3.105.0)
-    aws-sdk-kms (1.38.0)
+    aws-sdk-kms (1.37.0)
-    aws-sdk-s3 (1.81.1)
+    aws-sdk-s3 (1.79.0)
-    bcrypt (3.1.16)
-    better_errors (2.8.1)
+    bcrypt (3.1.13)
+    better_errors (2.7.1)
-    brakeman (4.10.0)
+    brakeman (4.9.0)
-    coderay (1.1.3)
+    coderay (1.1.2)
-    devise (4.7.3)
+    devise (4.7.2)
-    devise_invitable (2.0.2)
-      actionmailer (&gt;= 5.0)
-      devise (&gt;= 4.6)
-    excon (0.76.0)
-    flipper (0.19.0)
-    flipper-active_record (0.19.0)
+    flipper (0.18.0)
+    flipper-active_record (0.18.0)
-      flipper (~&gt; 0.19.0)
-    flipper-ui (0.19.0)
+      flipper (~&gt; 0.18.0)
+    flipper-ui (0.18.0)
-      flipper (~&gt; 0.19.0)
+      flipper (~&gt; 0.18.0)
-    jbuilder (2.10.1)
-      activesupport (&gt;= 5.0.0)
-    jeff (2.0.2)
-      excon
+    jbuilder (2.9.1)
+      activesupport (&gt;= 4.2.0)
-    loofah (2.7.0)
+    loofah (2.6.0)
-    minitest (5.14.2)
+    minitest (5.14.1)
-    nio4r (2.5.4)
+    nio4r (2.5.2)
-    omniauth-shopify-oauth2 (2.2.3)
+    omniauth-shopify-oauth2 (2.2.0)
-    peddler (2.4.3)
-      excon (&gt;= 0.50.0)
-      jeff (~&gt; 2.0)
-      multi_xml (&gt;= 0.5.0)
-    public_suffix (4.0.6)
-    puma (5.0.2)
+    public_suffix (4.0.5)
+    puma (4.3.5)
-    redis (4.2.2)
+    redis (4.2.1)
-    rollbar (3.0.0)
+    rollbar (2.27.0)
-    rubocop-ast (0.3.0)
-      parser (&gt;= 2.7.1.4)
-    rubocop-performance (1.8.0)
-      rubocop (&gt;= 0.87.0)
+    rubocop-ast (0.2.0)
+      parser (&gt;= 2.7.0.1)
+    rubocop-performance (1.7.1)
+      rubocop (&gt;= 0.82.0)
-    sendgrid-ruby (6.3.5)
+    sendgrid-ruby (6.3.4)
-    sidekiq (6.1.2)
+    sidekiq (6.1.1)
-    stripe (5.25.0)
+    stripe (5.24.0)
-    warden (1.2.9)
-      rack (&gt;= 2.0.9)
+    warden (1.2.8)
+      rack (&gt;= 2.0.6)
-    webmock (3.9.1)
+    webmock (3.8.2)
-    wkhtmltopdf-binary (0.12.6.4)
-    wkhtmltopdf-heroku (2.12.6.0)
+    wkhtmltopdf-binary (0.12.6.3)
+    wkhtmltopdf-heroku (2.12.5.0)
-  devise_invitable
-  jbuilder (~&gt; 2.10)
+  jbuilder (~&gt; 2.7)
-  peddler
-  puma (~&gt; 5.0)
+  puma (~&gt; 4.3)
-  wkhtmltopdf-heroku (= 2.12.6.0)
+  wkhtmltopdf-heroku (= 2.12.5.0)
</code></pre>

**Comment by IanVaughan** [comment]
<p>This has just started failing again, even on previously passing builds, all the tests pass locally.<br>
So there is something external to my app that is affecting us.</p>
<p>Having to run without feature tests on CI now, hoping that all the team run a complete test run before pushing!!</p>
<p><code>bundler exec rspec --tag ~js</code></p>
<p>Last time this fixed itself after a few days, so I guess whatever was changed on the runner or VM was either rolled back or updated?</p>
<p>workflow yml and runner log output: <a href="https://gist.github.com/IanVaughan/3add69d297d1f237f653ff8f65bdae54" class="inline-onebox" rel="noopener nofollow ugc">runner.log · GitHub</a></p>
<p>Also raised issue here <a href="https://github.com/actions/virtual-environments/issues/3548" class="inline-onebox" rel="noopener nofollow ugc">Rails feature tests have stopped running javascript feature tests · Issue #3548 · actions/virtual-environments · GitHub</a></p>

**Comment by IanVaughan** [comment, marked as answer]
<p>We found the issue, a cleaver colleague than me, dig into the logs and spotted 2 issues, the first I think being the cause.</p>
<ol>
<li>running out of inotify watchers:</li>
</ol>
<pre><code class="lang-auto">        FATAL: Listen error: unable to monitor directories for changes.
        Visit https://github.com/guard/listen/blob/master/README.md#increasing-the-amount-of-inotify-watchers for info on how to fix this.
</code></pre>
<p>1.1 The fix was to set <code>DISABLE_SPRING: true</code> in GitHub actions</p>
<ol start="2">
<li>syntax error in test.js</li>
</ol>
<pre><code class="lang-auto">Compiling...
Compilation failed:
/home/runner/work/org/repo/config/webpack/test.js:7
import $ from 'jquery';
</code></pre>
<p>2.1. <code>test.js</code> hasn’t changed for nearly a year, but JS libraries have been upgraded each week.<br>
2.2 The fix was:</p>
<pre><code class="lang-diff">-import $ from 'jquery';
+const $ = require('jquery');
</code></pre>

**Comment by mtomov** [comment]
<p>I can confirm that disabling Spring indeed fixed the issue. Thanks a lot <a class="mention" href="/ianvaughan">@IanVaughan</a>  for sharing! Something like this would’ve easily killed a few days.</p>
<pre><code class="lang-auto">- name: Run Ruby tests
  env:
    DISABLE_SPRING: true
  run: ....
</code></pre>
